com.unity.netcode.gameobjects/Components/NetworkAnimator.cs

```diff
@@ -1,79 +1,169 @@
+            public KeyValuePair<int, int>[] IntParamArray;
+
             public HashSet<int> TriggerParameters;
             public LayerState[] LayerStates;
+            private const int k_InvalidKey = -1;
 
-            public AnimatorSnapshot(Dictionary<int, bool> boolParameters, Dictionary<int, float> floatParameters, Dictionary<int, int> intParameters, HashSet<int> triggerParameters, LayerState[] layerStates)
+            public AnimatorSnapshot(int boolCount, int floatCount, int intCount, int triggerCount, int layerStatesCount)
             {
-                BoolParameters = boolParameters;
-                FloatParameters = floatParameters;
-                IntParameters = intParameters;
-                TriggerParameters = triggerParameters;
-                LayerStates = layerStates;
+                BoolParamArray = new KeyValuePair<int, bool>[boolCount];
+                IntParamArray = new KeyValuePair<int, int>[intCount];
+                FloatParamArray = new KeyValuePair<int, float>[floatCount];
+                TriggerParameters = new HashSet<int>(triggerCount);
+                LayerStates = new LayerState[layerStatesCount];
+
+                SetBuffersToDefaultValues();
             }
 
             public AnimatorSnapshot()
             {
-                BoolParameters = new Dictionary<int, bool>(0);
-                FloatParameters = new Dictionary<int, float>(0);
-                IntParameters = new Dictionary<int, int>(0);
-                TriggerParameters = new HashSet<int>();
+                BoolParamArray = new KeyValuePair<int, bool>[0];
+                IntParamArray = new KeyValuePair<int, int>[0];
+                FloatParamArray = new KeyValuePair<int, float>[0];
+                TriggerParameters = new HashSet<int>(0);
                 LayerStates = new LayerState[0];
+                SetBuffersToDefaultValues();
+            }
+
+            private void SetBuffersToDefaultValues()
+            {
+                for (int i = 0; i < BoolParamArray.Length; i++)
+                {
+                    BoolParamArray[i] = new KeyValuePair<int, bool>(k_InvalidKey, false);
+                }
+
+                for (int i = 0; i < FloatParamArray.Length; i++)
+                {
+                    FloatParamArray[i] = new KeyValuePair<int, float>(k_InvalidKey, 0);
+                }
+
+                for (int i = 0; i < IntParamArray.Length; i++)
+                {
+                    IntParamArray[i] = new KeyValuePair<int, int>(k_InvalidKey, 0);
+                }
+
+                TriggerParameters.Clear();
+
+                Array.Clear(LayerStates, 0, LayerStates.Length);
             }
 
             public bool SetInt(int key, int value)
             {
-                if (IntParameters.TryGetValue(key, out var existingValue) && existingValue == value)
+                bool setOrUpdatedValue = false;
+
+                int existingKvIndex = Array.FindIndex(IntParamArray, pair => pair.Key == key);
```

Comment:
Is there a way to avoid the O(n) lookup?

Reply:
I'll see if I can re-structure this so that we can have a dictionary-style lookup - I'll start by swapping out custom serialization for the array approach you suggested.